File: Dynamic-Programming/300-M-Longest-Increasing-Subsequence/longest_increasing_subsequence.py
class Solution:
    def lengthOfLIS(self, nums: list[int]) -> int:
        m = len(nums)
        OPT = [1] * m

        for i in range(m-1, -1, -1):
            for j in range(i+1, m):
                if nums[i] < nums[j]:
                    OPT[i] = max(OPT[i], 1 + OPT[j])
        return max(OPT)

File: Dynamic-Programming/300-M-Longest-Increasing-Subsequence/longest_increasing_subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int m = nums.size();
        vector<int> OPT(m, 1);
        int result = 1;

        for (int i=m-1; i>=0; i--) {
            for (int j=i+1; j<m; j++) {
                if (nums[i] < nums[j]) {
                    OPT[i] = max(OPT[i], 1 + OPT[j]);
                    if (OPT[i] > result) {
                        result = OPT[i];
                    }
                }
            }
        }
        return result;
    }
};